internal/fingerprint/task.go

```diff
@@ -0,0 +1,103 @@
+package fingerprint
+
+import (
+	"context"
+
+	"github.com/go-task/task/v3/internal/logger"
+	"github.com/go-task/task/v3/taskfile"
+)
+
+type (
+	CheckerOption func(*CheckerConfig)
+	CheckerConfig struct {
+		statusChecker  StatusCheckable
+		sourcesChecker SourcesCheckable
+	}
+)
+
+func WithStatusChecker(checker StatusCheckable) CheckerOption {
+	return func(config *CheckerConfig) {
+		config.statusChecker = checker
+	}
+}
+
+func WithSourcesChecker(checker SourcesCheckable) CheckerOption {
+	return func(config *CheckerConfig) {
+		config.sourcesChecker = checker
+	}
+}
+
+func IsTaskUpToDate(
+	ctx context.Context,
+	t *taskfile.Task,
+	method string,
+	tempDir string,
+	dry bool,
+	logger *logger.Logger,
+	opts ...CheckerOption,
+) (bool, error) {
+	var statusUpToDate bool
+	var sourcesUpToDate bool
+	var err error
+
+	// If the task method is set, override the default method
+	if t.Method != "" {
+		method = t.Method
+	}
```

Comment:
Yeah, I don't love this either. Originally, I moved the method overriding into the function as I thought it would nice to not repeat it in 3/4 places, but it actually makes far more sense to check on the caller side.

It might be nice to set the task's method to the default specified in the Taskfile if it isn't overridden when reading/compiling? This way, we wouldn't have to do this check at all.

WDYT? and do you think this should this be done when the file is read or when it is compiled?